### PR TITLE
Fix calculated fields bugs

### LINKED ISSUE #260: Boolean calculated fields eval to true when they should be false
Example: when fields A, B and C are all `0`, a calculated field with expression `A or B or C` will still return `true`.

Reason: We're not casting the CSV column values to the correct types before evaluating the expression, so the expression evaluator is comparing the string `"0"` (which is `true`) rather than the number `0` (which is - in our expression evaluator, anyway - `false`)

### PR DESCRIPTION
**Pull request by yozlet**
Fixes #260 & #261.

### CODE REVIEW

File: lib/data_magic/config.rb [hunk 1]
```diff
@@ -177,6 +181,21 @@ def column_field_types
       @column_types
     end
 
+    # actually does what column_field_types is meant to do - creates a hash of CSV column
```
Comment (ultrasaurus) [inline lib/data_magic/config.rb]: column_field_types does what it is meant to do... it's the field name for the column mapped to the type.... just in retrospect, it is easier to reason about the code if we just use column names OR full field names and omit the weird intermediary format.

File: lib/expression/eval.rb [hunk 2]
```diff
@@ -7,11 +7,13 @@ class Eval < Parslet::Transform
 
     # in Ruby 0 is 'truthy' but that's not what most people expect
     rule(:or => { :left => subtree(:left), :right => subtree(:right) }) do
-      left == 0 ? right : (left or right)
+      result = left == 0 ? right : (left or right)
+      result.nil? ? 0 : result
```
Comment (ultrasaurus) [inline lib/expression/eval.rb]: if we OR two nil values, don't we want the result to be nil...?  this seems to transform it into 0. not sure why we would do that

File: lib/expression/eval.rb [hunk 3]
```diff
@@ -7,11 +7,13 @@ class Eval < Parslet::Transform
...
     end
 
     rule(:and => { :left => subtree(:left), :right => subtree(:right) }) do
-      left == 0 ? left : (left and right)
+      result = left == 0 ? left : (left and right)
+      result.nil? ? 0 : result
```
Comment (ultrasaurus) [inline lib/expression/eval.rb]: same question about nil as previous comment

File: spec/lib/expression_spec.rb [hunk 4]
```diff
@@ -6,25 +6,47 @@
...
       expect(Expression.new(expr).evaluate(values)).to eq(1)
     end
+
+    it "evaluates: 1 OR nil to be 1" do
+      expr = "t1 or t2"
+      values = {t1:1, t2:nil}
+      expect(Expression.new(expr).evaluate(values)).to eq(1)
+    end
+
+    it "evaluates: 0 OR nil to be 0" do
```
Comment (ultrasaurus) [inline spec/lib/expression_spec.rb]: 0 or nil should be nil, I think

File: spec/lib/expression_spec.rb [hunk 5]
```diff
@@ -6,25 +6,47 @@
...
+      expr = "t1 or t2"
+      values = {t1:0, t2:nil}
+      expect(Expression.new(expr).evaluate(values)).to eq(0)
+    end
+
+    it "evaluates: nil OR nil to be 0" do
```
Comment (ultrasaurus) [inline spec/lib/expression_spec.rb]: ?

File: spec/lib/expression_spec.rb (hunk 4 above)
Comment (yozlet) [inline spec/lib/expression_spec.rb]: Post-discussion: yep, this is what we've agreed on (follow same OR ordering as JS/Ruby)

**Comment by yozlet** [comment]
OK, after discussion with @ultrasaurus, we decided to go with following programming order of OR evaluation. So:
 * `0 OR nil == nil`
 * `nil OR 0 == 0`
 * `nil OR nil == nil`

**Comment by ultrasaurus** [comment]
Note: this is consistent with Ruby and Javascript evaluation of OR expressions

**Comment by ultrasaurus** [comment]
:+1:

**Pull request merged**